dal/elastic.go

```diff
@@ -2,126 +2,109 @@ package dal
 	ctx := context.Background()
-	//get document count in index
-	countService := elastic.NewCountService(client)
-	countResult, err := countService.Index(INDEXNAME).Do(ctx)
+	putStatus, err := (e.Client).Index().Index(INDEX_NAME).BodyJson(*book).Do(ctx)
 	if err != nil {
-		panic(err)
-	}
-
-	//get number of distinct authors
-	agg := elastic.NewCardinalityAggregation().Field("author_name.keyword")
-	diffAuthors, err := client.Search().Index(INDEXNAME).Aggregation("diff_authors", agg).Do(ctx)
-	if err !=nil{
-		panic(err)
+		return "", err
 	}
-	return countResult,diffAuthors
-
+	return putStatus.Id, nil
 }
-
-func ElasticPutBook(client *elastic.Client, book *Book) *elastic.IndexResponse{
+func (e *Elastic) UpdateBook(title string, id string) error {
 	ctx := context.Background()
-	// Index a tweet (using JSON serialization)
-	put1, err := client.Index().Index(INDEXNAME).BodyJson(*book).Do(ctx)
+	_, err := (e.Client).Update().Index(INDEX_NAME).Id(id).Doc(map[string]interface{}{"title": title}).Do(ctx)
 	if err != nil {
-		// Handle error
-		panic(err)
+		return err
 	}
-	return put1
+	return nil
 }
-
-func ElasticUpdateBook(client *elastic.Client, title string, id string) *elastic.UpdateResponse{
+func (e *Elastic) DeleteBook(id string) error {
 	ctx := context.Background()
-	response, err := client.Update().Index(INDEXNAME).Id(id).Doc(map[string]interface{}{"title": title}).Do(ctx)
+	_, err := (e.Client).Delete().Index("books_yuval").Id(id).Do(ctx)
 	if err != nil {
-		// Handle error
-		panic(err)
+		return err
 	}
-	return response
+	return nil
 }
 
-func ElasticDeleteBook(client *elastic.Client, id string) *elastic.DeleteResponse{
+func (e *Elastic) SearchBook(title string, authorName string, priceRangeStr string) ([]models.Book, error) {
 	ctx := context.Background()
-	response, err := client.Delete().Index("books_yuval").Id(id).Do(ctx)
+	query := elastic.NewBoolQuery()
+	if FieldExist(title) {
+		matchQuery1 := elastic.NewMatchQuery("title", title)
+		query.Filter(matchQuery1)
+	}
+	if FieldExist(authorName) {
+		matchQuery2 := elastic.NewMatchQuery("author_name", authorName)
+		query.Filter(matchQuery2)
+	}
+	if FieldExist(priceRangeStr) {
+		priceRange := GetPriceRange(priceRangeStr)
+		if priceRange != nil {
+			rangeQuery := elastic.NewRangeQuery("price").From(priceRange[0]).To(priceRange[1])
+			query.Filter(rangeQuery)
+		}
+	}
+	searchResult, err := (e.Client).Search().Index(INDEX_NAME).Query(query).Size(100).Do(ctx)
 	if err != nil {
-		// Handle error
-		panic(err)
+		return nil, err
 	}
-	return response
+	numHits := searchResult.Hits.TotalHits.Value
+	resultsArr := make([]models.Book, numHits)
+	for i, hit := range searchResult.Hits.Hits {
+		json.Unmarshal(hit.Source, &resultsArr[i])
+	}
+	return resultsArr, nil
 }
 
-func ElasticSearchBook(client *elastic.Client, title string, author_name string, price_range [2]float64, opcode int) *elastic.SearchResult{
+func (e *Elastic) GetStoreInfo() (int64, *float64, error) {
```

Comment:
If we will want to return more information  about our book store- you will need to change the signature of the function - think of a more convenient way to perform this task

Reply:
I feel that using a struct to return two numbers is too much, Especially since if we would want to return more information we would have to refactor the code here anyway. If we would want to return a new field we would have to implement elastic querying it, its not as if we just add the field to the struct and for example "Json.unmarshall" it in the function body so we dont have to refactor anything.
Did you mean using a struct? what do you think?